### PR TITLE
IGNITE-7609 .NET: FieldsQueryCursor should expose data types too

### PR DESCRIPTION
**Pull request by gurustron**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java [hunk 1]
```diff
@@ -70,6 +84,17 @@ public PlatformFieldsQueryCursor(PlatformContext platformCtx, QueryCursorEx<List
             for (int i = 0; i < cnt; i++) {
                 writer.writeString(fq.getFieldName(i));
             }
+        } else if (type == OP_GET_FIELDS_META) {
+            QueryCursorEx<List<?>> cursor = cursor();
+
+            List<GridQueryFieldMetadata> metadatas = cursor.fieldsMeta();
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java:90]: `metadatas` -> `metas`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java [hunk 2]
```diff
@@ -70,6 +84,17 @@ public PlatformFieldsQueryCursor(PlatformContext platformCtx, QueryCursorEx<List
...
+
+            writer.writeInt(metadatas.size());
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java:92]: `cursor.fieldsMeta()` is not guaranteed to return a non-null result, let's add a check.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java [hunk 3]
```diff
@@ -70,6 +84,17 @@ public PlatformFieldsQueryCursor(PlatformContext platformCtx, QueryCursorEx<List
...
+
+            for (GridQueryFieldMetadata metadata: metadatas){
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java:94]: `metadata` -> `meta`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java:90]: https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules

Weird rules, but we follow them.

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IFieldsQueryCursor.cs [hunk 4]
```diff
@@ -30,5 +31,10 @@ public interface IFieldsQueryCursor : IQueryCursor<IList<object>>
         /// Gets the field names.
         /// </summary>
         IList<string> FieldNames { get; }
+
+        /// <summary>
+        /// Gets fields metadata
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IFieldsQueryCursor.cs:36]: Comment should end with `.`

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs [hunk 5]
```diff
@@ -33,7 +34,7 @@ internal class FieldsQueryCursor<T> : PlatformQueryQursorBase<T>
         /// Constructor.
         /// </summary>
         /// <param name="target">Target.</param>
-        /// <param name="keepBinary">Keep poratble flag.</param>
+        /// <param name="keepBinary">Keep portable flag.</param>
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs:37]: `portable` -> `binary`.

P.S. This is an ancient legacy of early Ignite days. `Binary` was called `Portable`, then we renamed, but this one line was skipped due to a typo.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs [hunk 6]
```diff
@@ -85,5 +92,24 @@ public IList<string> FieldNames
                            Target.OutStream(OpGetFieldNames, reader => reader.ReadStringCollection())));
             }
         }
+
+        public IList<IQueryCursorFieldMetadata> FieldsMetadata
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs:96]: Missing XMLDoc

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IFieldsQueryCursor.cs [hunk 7]
```diff
@@ -30,5 +31,10 @@ public interface IFieldsQueryCursor : IQueryCursor<IList<object>>
...
+        /// </summary>
+        IList<IQueryCursorFieldMetadata> FieldsMetadata { get; }
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IFieldsQueryCursor.cs:38]: Let's just name this `Fields`

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Apache.Ignite.Core.Cache.Query
+{
+    using System;
+
+    /// <summary>
+    /// Query field descriptor. This descriptor is used to provide metadata
+    /// about fields returned in query result.
+    /// </summary>
+    public interface IQueryCursorFieldMetadata
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs:26]: In line with the above: `IQueryCursorField`

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        /// <summary>
+        /// Field name
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs:29]: `Gets the field name.`
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1623.md

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        /// Name of java type to which this field belongs
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs:34]: `Gets the name of Java type for this field`.

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        string JavaTypeName { get; }
+
+        /// <summary>
+        /// Type to which this field belongs
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IQueryCursorFieldMetadata.cs:39]: `Gets the field type`.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/QueryCursorFieldMetadataImpl.cs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+{
+    using System;
+    using Apache.Ignite.Core.Binary;
+    using Apache.Ignite.Core.Cache.Query;
+    using Apache.Ignite.Core.Impl.Binary;
+
+    /// <summary>
+    /// Query cursor field metadata implementation.
+    /// </summary>
+    public class QueryCursorFieldMetadataImpl : IQueryCursorFieldMetadata
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/QueryCursorFieldMetadataImpl.cs:28]: Please remove `Impl` suffix. It is present on some types because of Java naming legacy, but it is against Microsoft standards, so we try to avoid this.

Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/QueryCursorFieldMetadataImpl.cs:28]: **Important**: make this class `internal`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java [hunk 13]
```diff
@@ -22,29 +22,40 @@
...
+import org.apache.ignite.internal.processors.query.GridQueryFieldMetadata;
 
 import java.util.List;
 
 /**
  * Interop cursor for fields query.
  */
 public class PlatformFieldsQueryCursor extends PlatformAbstractQueryCursor<List<?>> {
-    /** Gets field names. */
+    /**
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java:33]: Something happened with comments blocks here and below. Please revert unrelated changes.

File: modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IFieldsQueryCursor.cs [hunk 14]
```diff
@@ -19,6 +19,7 @@ namespace Apache.Ignite.Core.Cache.Query
 {
     using System.Collections.Generic;
     using System.Diagnostics.CodeAnalysis;
+    using Apache.Ignite.Core.Impl.Cache.Query;
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Cache/Query/IFieldsQueryCursor.cs:22]: Unused import?

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs [hunk 15]
```diff
@@ -85,5 +92,26 @@ public IList<string> FieldNames
...
+                if (_fieldsMeta == null)
+                {
+                    var metadata = Target.OutStream(
+                        OpGetFieldsMeta,
+                        reader => reader.ReadCollectionRaw(stream =>
+                            new QueryCursorField(stream) as IQueryCursorField))
+                        ?? new List<IQueryCursorField>();
+
+                    _fieldsMeta = new ReadOnlyCollection<IQueryCursorField>(
+                        metadata as List<IQueryCursorField> ?? metadata.ToList());
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs:110]: `ToList` causes unnecessary allocations here. My proposal:

 * Remove `??` above, get the result of `ReadCollectionRaw` as is
* `_fieldsMeta = meta != null ? meta.AsReadOnly() : new ReadOnlyCollection()`

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/QueryCursorField.cs [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+        /// <summary>
+        ///  Gets .NET type that corresponds to specified Java type name.
+        /// </summary>
+        /// <param name="javaTypeName">Name of the java type.</param>
+        /// <returns></returns>
+        private static Type GetDotNetType(string javaTypeName)
+        {
+            var dotNetType = JavaTypes.GetDotNetType(javaTypeName);
+            if (dotNetType == null && javaTypeName == "java.lang.Object") dotNetType = typeof(object);
```
Comment (ptupitsyn) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/QueryCursorField.cs:58]: Can we move this check to `JavaTypes` class?

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs (hunk 15 above)
Comment (gurustron) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs:110]: As far as i understand .AsReadOnly() is not an option here cause we have IList in contract and if the underlying collection is mutable it will allow mutations.

Comment (gurustron) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Query/FieldsQueryCursor.cs]: Changed the underlying method to return IList since it is already returning one

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java (hunk 13 above)
Comment (gurustron) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformFieldsQueryCursor.java]: Auto-formatting by IDE.
